### PR TITLE
[Auto Parallel] Support semi-auto trainer and fit Llama2 training

### PR DESCRIPTION
**Pull request by haohongxiang**
### PR types

Bug fixes
### PR changes

Others
### Description

[Auto Parallel] Support semi-auto trainer and fit Llama2 training

### CODE REVIEW

**Review by ZHUI — CHANGES_REQUESTED** [review]
改动比较大，先 Request changes 一手。

File: llm/llama/auto_parallel/run_pretrain_3D_auto.py [hunk 1]
```diff
@@ -314,29 +362,17 @@ def get_train_data_file(args):
...
-        parameters=model.parameters(),
-        weight_decay=training_args.weight_decay,
-        grad_clip=paddle.nn.ClipGradByGlobalNorm(training_args.max_grad_norm)
-        if training_args.max_grad_norm > 0
-        else None,
-        **optimizer_kwargs,
-    )
-
-    return optimizer
+    def _wrap_dist_loader(self, train_dataloader):
```
Comment (zhiqiu) [inline llm/llama/auto_parallel/run_pretrain_3D_auto.py]: It is not used in dynamic mode?

File: paddlenlp/trainer/auto_trainer.py [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+            if "pp" in mesh.dim_names:
+                mesh = mesh.get_mesh_with_dim("pp")[pp_idx]
+            return mesh
+
+        meshes = []
+        for pp_idx in range(self.args.pipeline_parallel_degree):
+            meshes.append(_get_mesh(pp_idx))
+        return meshes
+
+    def _wrap_dist_loader(self, train_dataloader):
```
Comment (zhiqiu) [inline paddlenlp/trainer/auto_trainer.py]: what's the difference with `_wrap_dist_loader` in run_pretrain_3D_auto.py?

File: paddlenlp/trainer/auto_trainer.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+        return dist.shard_dataloader(
+            dataloader=train_dataloader,
+            meshes=self._get_meshes_for_loader(),
+            shard_dims="dp",
+        )
+
+    def _wrap_for_static(self, model, train_dataloader):
```
Comment (zhiqiu) [inline paddlenlp/trainer/auto_trainer.py]: it seems not used?

File: paddlenlp/transformers/llama/modeling_3D_auto.py [hunk 4]
```diff
@@ -939,15 +877,16 @@ def forward(
         if position_ids is None:
             position_ids = paddle.arange(seq_length, dtype="int64").expand((batch_size, seq_length))
             # NOTE(zhaoyingli): infer spmd does not support [seq_len] --> [batch, seq_len] in data_parallel
-            position_ids = dist.shard_tensor(position_ids, get_mesh(), [dist.Shard(0), dist.Replicate()])
+            position_ids = dist.shard_tensor(position_ids, get_mesh(), [dist.Replicate(), dist.Replicate()])
```
Comment (zhiqiu) [inline paddlenlp/transformers/llama/modeling_3D_auto.py:912]: why change to replicated?

File: llm/llama/auto_parallel/run_pretrain_3D_auto.py (hunk 1 above)
Comment (haohongxiang) [reply llm/llama/auto_parallel/run_pretrain_3D_auto.py]: Done, now it's used in dynamic and static mode.

File: paddlenlp/trainer/auto_trainer.py (hunk 3 above)
Comment (haohongxiang) [reply paddlenlp/trainer/auto_trainer.py]: It's called in `Trainer` from `paddlenlp/trainer/trainer.py`, for wrapping `model` into `DistModel` in static mode

File: paddlenlp/transformers/llama/modeling_3D_auto.py (hunk 4 above)
Comment (haohongxiang) [reply paddlenlp/transformers/llama/modeling_3D_auto.py:912]: Because in static mode, infer spmd hasn't supported the case -- "[seq_len] --> [batch, seq_len]"

File: paddlenlp/trainer/trainer.py [hunk 5]
```diff
@@ -718,6 +720,11 @@ def train(
                 self.create_optimizer_and_scheduler(num_training_steps=max_steps)
             self._load_optimizer_and_scheduler(resume_from_checkpoint)
 
+        if self.args.use_auto_parallel and self.args.run_static_semi_auto:
+            model = self._wrap_for_static(model, train_dataloader)
+
+        self.model = model
```
Comment (ZHUI) [inline paddlenlp/trainer/trainer.py]:
```suggestion
        if self.args.use_auto_parallel and self.args.run_static_semi_auto:
            model = self._wrap_for_static(model, train_dataloader)
            self.model = model
```

File: paddlenlp/trainer/trainer.py [hunk 6]
```diff
@@ -828,7 +818,8 @@ def train(
 
         self.control = self.callback_handler.on_train_begin(args, self.state, self.control)
 
-        tr_loss = paddle.to_tensor(0.0)
+        with _exec_mode_guard("dynamic"):
+            tr_loss = paddle.to_tensor(0.0)
```
Comment (ZHUI) [inline paddlenlp/trainer/trainer.py]: 这个能不改吗？以后是不是加一些这种直接操作要很小心了？

Comment (ZHUI) [reply paddlenlp/trainer/trainer.py]: 要不我给你们把这个主loop解耦？

File: paddlenlp/trainer/trainer.py (hunk 5 above)
Comment (haohongxiang) [reply paddlenlp/trainer/trainer.py]: Done

File: paddlenlp/trainer/trainer.py (hunk 6 above)
Comment (haohongxiang) [reply paddlenlp/trainer/trainer.py]: Done，直接重写`train`函数，已避免这类问题

File: paddlenlp/trainer/trainer.py [hunk 7]
```diff
@@ -682,6 +659,56 @@ def train(
...
             )
 
+        logger.info("***** Running training *****")
+        logger.info(f"  Num examples = {num_examples:,}")
+        logger.info(f"  Num Epochs = {num_train_epochs}")
+        logger.info(f"  Instantaneous batch size per device = {args.per_device_train_batch_size}")
+        logger.info(f"  Total train batch size (w. parallel, distributed & accumulation) = {total_train_batch_size}")
+        logger.info(f"  Gradient Accumulation steps = {args.gradient_accumulation_steps}")
+        logger.info(f"  Total optimization steps = {max_steps:,}")
+        logger.info(f"  Total num train samples = {num_train_samples:,}")
```
Comment (ZHUI) [inline paddlenlp/trainer/trainer.py]: 这里打印的话，但是没有看到主loop那里的打印删除

File: paddlenlp/trainer/trainer.py [hunk 8]
```diff
@@ -601,51 +602,27 @@ def _wrap_model_and_load_sharded_checkpoint(self, resume_from_checkpoint):
...
-                `bool` and equals `True`, load the last checkpoint in *args.output_dir* as saved by a previous instance
-                of [`Trainer`]. If present, training will resume from the model/optimizer/scheduler states loaded here.
-            ignore_keys_for_eval (`List[str]`, *optional*)
-                A list of keys in the output of your model (if it is a dictionary) that should be ignored when
-                gathering predictions for evaluation during the training.
-        """
-        args = self.args
-        self.is_in_train = True
-
+    def _sync_resume_states(self, resume_from_checkpoint):
```
Comment (ZHUI) [inline paddlenlp/trainer/trainer.py]: 这个函数我看你的 auto_trainer 是复用的，是不是可以留着？

File: paddlenlp/trainer/trainer.py [hunk 9]
```diff
@@ -1169,13 +1196,17 @@ def _print_timer(self):
...
+        return loss.item()
+
     def _maybe_log_save_evaluate(self, tr_loss, model, epoch, ignore_keys_for_eval, **kwargs):
         if self.control.should_log:
 
             logs: Dict[str, float] = {}
 
             # all_gather + mean() to get average loss over all processes
-            tr_loss_scalar = self._nested_gather(tr_loss).mean().item()
+            tr_loss_scalar = self._get_item_from_loss(self._nested_gather(tr_loss).mean())
```
Comment (ZHUI) [inline paddlenlp/trainer/trainer.py:1220]: https://github.com/PaddlePaddle/PaddleNLP/blob/fe6b45d8666e466da6c4ebb6f5c64e9baebd7e4b/paddlenlp/trainer/trainer.py#L1199-L1209

这里 我看你是复用了 _maybe_log_save_evaluate 函数。而且外面包括了 guard，为什么这里 要加一个 `assert isinstance(loss, paddle.Tensor) and loss._is_initialized()`的检查？

File: paddlenlp/trainer/training_args.py [hunk 10]
```diff
@@ -747,6 +748,8 @@ class TrainingArguments:
         default=False,
         metadata={"help": "reshard pp even if pp degree in the model and pp degree in script match"},
     )
+    parallel_mode: str = field(default="hybrid", metadata={"help": ""})
```
Comment (ZHUI) [inline paddlenlp/trainer/training_args.py]: 这里注释写详细一点？标注一下只为自动并行或者半自动并行使用？有什么是选的？

File: paddlenlp/trainer/training_args.py [hunk 11]
```diff
@@ -747,6 +748,8 @@ class TrainingArguments:
...
+    run_static_semi_auto: bool = field(default=True, metadata={"help": ""})
```
Comment (ZHUI) [inline paddlenlp/trainer/training_args.py]: 这个参数的具体意义又是？两个选项是否存在合并的可能？

File: llm/llama/auto_parallel/run_pretrain_3D_auto.py [hunk 12]
```diff
@@ -18,32 +18,34 @@
...
 import numpy as np
 import paddle
 import paddle.distributed as dist
 from paddle.distributed import fleet
-from paddle.io import DataLoader, DistributedBatchSampler
 
-from paddlenlp.trainer import PdArgumentParser, Trainer, TrainingArguments
+from paddlenlp.ops import Topology
+from paddlenlp.trainer import PdArgumentParser, TrainingArguments, get_last_checkpoint
+from paddlenlp.trainer.auto_trainer import SemiAutoTrainer
```
Comment (ZHUI) [inline llm/llama/auto_parallel/run_pretrain_3D_auto.py]: 自动并行这块具体是规划出来，只有这一个 SemiAutoTrainer 吗？
或者后面可能还会有 AutoTrainer?

File: paddlenlp/trainer/auto_trainer.py [hunk 13]
```diff
@@ -0,0 +1,597 @@
...
+    has_length,
+    speed_metrics,
+)
+from .utils.helper import distributed_file, distributed_isfile  # nested_truncate,
+
+DEFAULT_CALLBACKS = [DefaultFlowCallback]
+DEFAULT_PROGRESS_CALLBACK = ProgressCallback
+
+# Name of the files used for checkpointing
+TRAINING_ARGS_NAME = "training_args.bin"
```
Comment (ZHUI) [inline paddlenlp/trainer/auto_trainer.py]: 这些个name trainer.py 里面有的话，改成import trainer.py 里面的吧

File: paddlenlp/trainer/auto_trainer.py [hunk 14]
```diff
@@ -0,0 +1,597 @@
...
+SCALER_NAME = "scaler.pdparams"
+
+
+class SemiAutoTrainer(Trainer):
+    def __init__(self, *args, **kwargs):
+
+        if kwargs.get("args", None) is not None and kwargs["args"].run_static_semi_auto:
+            if kwargs.get("criterion", None) is None:
+
+                def loss_func(loss, outputs):
```
Comment (ZHUI) [inline paddlenlp/trainer/auto_trainer.py:54]: 需要 定义  criterion 吗？现在paddlenlp的模型，loss基本在模型里面了。不额外定义。

File: paddlenlp/trainer/auto_trainer.py [hunk 15]
```diff
@@ -0,0 +1,597 @@
...
+        return TrainOutput(self.state.global_step, train_loss, metrics)
+
+    def _get_train_sampler(self) -> Optional[paddle.io.Sampler]:
+        if self.train_dataset is None or not has_length(self.train_dataset):
+            return None
+
+        total_batch_size_per_acc_step = self.args.per_device_train_batch_size * self.args.dataset_world_size
+        total_batch_size = total_batch_size_per_acc_step * self.args.gradient_accumulation_steps
+        batch_size = total_batch_size if self.args.run_static_semi_auto else total_batch_size_per_acc_step
+
```
Comment (ZHUI) [inline paddlenlp/trainer/auto_trainer.py:368]: 不是很懂，是说 acc 会并 run_static_semi_auto 里面控制，所以bs更大？

会不会出现数据不一致问题，开启**run_static_semi_auto** 与否？

File: paddlenlp/trainer/trainer.py (hunk 9 above)
Comment (haohongxiang) [reply paddlenlp/trainer/trainer.py:1220]: 这里可以删掉，半自动判断逻辑在 `auto_trainer` 中重写即可

File: paddlenlp/trainer/auto_trainer.py (hunk 13 above)
Comment (haohongxiang) [reply paddlenlp/trainer/auto_trainer.py]: Done

File: llm/llama/auto_parallel/run_pretrain_3D_auto.py (hunk 12 above)
Comment (haohongxiang) [reply llm/llama/auto_parallel/run_pretrain_3D_auto.py]: 目前规划只有SemiAutoTrainer；未来推全自动架构的话，可以在当前这套trainer架构里做升级

File: paddlenlp/trainer/training_args.py (hunk 11 above)
Comment (haohongxiang) [reply paddlenlp/trainer/training_args.py]: 这个参数用以区分动半执行还是静半执行；默认值为`True`，表示在静半模型下执行训练端到端流程；若手动设置为`False`，在动半模式下执行训练，可方便用户完成组网标记等模块的调试。

File: paddlenlp/trainer/auto_trainer.py (hunk 15 above)
Comment (haohongxiang) [reply paddlenlp/trainer/auto_trainer.py:368]: 已修改，batch sample里统一传入`global batch size`；除静半PP策略，其他场景的梯度累加都需要在读出`global batch`后在`batch dim`维度做split，然后for循环执行，完成梯度累加

File: paddlenlp/trainer/auto_trainer.py (hunk 14 above)
Comment (haohongxiang) [reply paddlenlp/trainer/auto_trainer.py:54]: 静半架构运行需要一个假的critirion，哪怕直接返回loss也可以

File: llm/llama/auto_parallel/run_pretrain_3D_auto.py (hunk 12 above)
Comment (haohongxiang) [reply llm/llama/auto_parallel/run_pretrain_3D_auto.py]: 目前已更名为`AutoTrainer`

File: paddlenlp/trainer/training_args.py (hunk 10 above)
Comment (haohongxiang) [reply paddlenlp/trainer/training_args.py]: Done

File: paddlenlp/trainer/training_args.py [hunk 16]
```diff
@@ -747,6 +748,18 @@ class TrainingArguments:
...
+        metadata={
+            "help": (
+                "Which parallel mode to use for distributed training.\n"
+                "Following options are supports:\n"
+                "- hybrid: under the hybrid parallel mode with combined distributed strategies.\n"
+                "- auto: under the auto parallel mode with AutoTrainer \n"
+            )
+        },
+    )
+    run_static_auto: bool = field(default=True, metadata={"help": "whether to run static graph in auto parallel mode"})
```
Comment (ZHUI) [inline paddlenlp/trainer/training_args.py]: How about?
```python
"hybrid"
"auto"
"auto_static"
"auto_semi"
"auto_semi_static"
```

File: paddlenlp/trainer/trainer.py (hunk 9 above)
Comment (ZHUI) [reply paddlenlp/trainer/trainer.py:1220]: 辛苦删除一下？

**Review by ZHUI — APPROVED** [review]
LGTM
